libraries/core_libs/consensus/src/dag/sortition_params_manager.cpp

```diff
@@ -0,0 +1,176 @@
+#include "dag/sortition_params_manager.hpp"
+
+#include "pbft/pbft_block.hpp"
+namespace taraxa {
+
+SortitionParamsChange::SortitionParamsChange(uint64_t period, uint16_t efficiency, VrfParams const& vrf,
+                                             SortitionParamsChange const& previous)
+    : period(period), vrf_params(vrf), interval_efficiency(efficiency) {
+  actual_correction_per_percent = correctionPerPercent(previous);
+}
+
+SortitionParamsChange::SortitionParamsChange(uint64_t period, uint16_t efficiency, VrfParams const& vrf)
+    : period(period), vrf_params(vrf), interval_efficiency(efficiency) {}
+
+uint16_t SortitionParamsChange::correctionPerPercent(SortitionParamsChange const& previous) const {
+  // make calculations on vrf upper threshold
+  const auto param_change = vrf_params.threshold_upper - previous.vrf_params.threshold_upper;
+  int32_t actual_efficiency_change = interval_efficiency - previous.interval_efficiency;
+  // to avoid division by zero
+  if (actual_efficiency_change == 0) {
+    actual_efficiency_change = 1;
+  }
+
+  auto correction_per_percent = std::abs(param_change * ONE_PERCENT / actual_efficiency_change);
+  // zero correction could lead to stuck of that protocol and give up with zero corrections
+  if (correction_per_percent == 0) {
+    correction_per_percent = 1;
+  }
+  // correction per percent shouldn't be bigger than percent of value
+  correction_per_percent = std::min(correction_per_percent, std::numeric_limits<uint16_t>::max() / 100);
+  return correction_per_percent;
+}
+
+bytes SortitionParamsChange::rlp() const {
+  dev::RLPStream s;
+  s.appendList(5);
+  s << vrf_params.threshold_lower;
+  s << vrf_params.threshold_upper;
+  s << period;
+  s << interval_efficiency;
+  s << actual_correction_per_percent;
+
+  return s.out();
+}
+
+SortitionParamsChange SortitionParamsChange::from_rlp(dev::RLP const& rlp) {
+  SortitionParamsChange p;
+
+  p.vrf_params.threshold_lower = rlp[0].toInt<uint16_t>();
+  p.vrf_params.threshold_upper = rlp[1].toInt<uint16_t>();
+  p.period = rlp[2].toInt<uint64_t>();
+  p.interval_efficiency = rlp[3].toInt<uint16_t>();
+  p.actual_correction_per_percent = rlp[4].toInt<uint16_t>();
+
+  return p;
+}
+
+SortitionParamsManager::SortitionParamsManager(addr_t node_addr, SortitionConfig sort_conf,
+                                               std::shared_ptr<DbStorage> db)
+    : config_(sort_conf), db_(db) {
+  LOG_OBJECTS_CREATE("SORTITION_MGR");
+  // load cache values from db
+  params_changes_ = db_->getLastSortitionParams(config_.changes_count_for_average);
+  // restore VRF params from last change
+  if (!params_changes_.empty()) {
+    config_.vrf = params_changes_.back().vrf_params;
+  }
+
+  dag_efficiencies_.reserve(config_.computation_interval);
+  dag_efficiencies_ = db_->getLastIntervalEfficiencies(config_.computation_interval);
+}
+uint64_t SortitionParamsManager::currentProposalPeriod() const {
+  if (params_changes_.empty()) return 0;
```

Comment:
I guess we should decide if we want to use parentheses for 1-line if's or not

Reply:
I am ok with this parentheses. But I think for this kind of simple condition checks(just one condition inside and simple return or value setting) we could have it in one line because it is pretty readable. 
But from the other side this rule seems to be very complicated

Reply:
But I think we should just use rules from the styleguide about it and do not create a new exceptions.
Because if we will introduce some exceptions to it - it will be a new style and we should kind support it like create a doc with a description and ideally change clang-format rules and I dont think it is worth it, etc

Reply:
https://google.github.io/styleguide/cppguide.html#Conditionals

- For historical reasons, we allow one exception to the above rules: if an if statement has no else or else if clauses, then the curly braces for the controlled statement or the line breaks inside the curly braces may be omitted

Well official google style guide allows not to use braces for historical reasons... Seems to me that there was a lot of legacy code with such `if's` so they created this exception. Anyway "always use braces" is just bulletproof rule so I personally would vote for it :)